x-pack/filebeat/input/salesforce/input.go

```diff
@@ -0,0 +1,506 @@
+		Name:      inputName,
+		Stability: feature.Stable,
+		Manager:   NewInputManager(log, store),
+	}
+}
+
+func (s *salesforceInput) Name() string { return inputName }
+
+func (s *salesforceInput) Test(_ inputcursor.Source, _ v2.TestContext) error {
+	return nil
+}
+
+// Run starts the input and blocks until it ends completes. It will return on
+// context cancellation or type invalidity errors, any other error will be retried.
+func (s *salesforceInput) Run(env v2.Context, src inputcursor.Source, cursor inputcursor.Cursor, pub inputcursor.Publisher) (err error) {
+	st := &state{}
+	if !cursor.IsNew() {
+		if err = cursor.Unpack(&st); err != nil {
+			return err
+		}
+	}
+
+	if err = s.Setup(env, src, st, pub); err != nil {
+		return err
+	}
+
+	return s.run()
+}
+
+// Setup sets up the input. It will create a new SOQL resource and all other
+// necessary configurations.
+func (s *salesforceInput) Setup(env v2.Context, src inputcursor.Source, cursor *state, pub inputcursor.Publisher) (err error) {
+	cfg := src.(*source).cfg
+
+	ctx := ctxtool.FromCanceller(env.Cancelation)
+	childCtx, cancel := context.WithCancelCause(ctx)
+
+	s.srcConfig = &cfg
+	s.ctx = childCtx
+	s.cancel = cancel
+	s.publisher = pub
+	s.cursor = cursor
+	s.log = env.Logger.With("input_url", cfg.URL)
+	s.sfdcConfig, err = getSFDCConfig(&cfg)
+	if err != nil {
+		return fmt.Errorf("error with configuration: %w", err)
+	}
+
+	s.soqlr, err = s.SetupSFClientConnection() // create a new SOQL resource
+	if err != nil {
+		return fmt.Errorf("error setting up connection to Salesforce: %w", err)
+	}
+
+	return nil
+}
+
+// run is the main loop of the input. It will run until the context is cancelled
+// and based on the configuration, it will run the different methods -- EventLogFile
+// or Object to collect events at defined intervals.
+func (s *salesforceInput) run() error {
+	if s.srcConfig.EventMonitoringMethod.EventLogFile.isEnabled() {
+		err := s.RunEventLogFile()
+		if err != nil {
+			s.log.Errorf("Problem running EventLogFile collection: %s", err)
+		}
+	}
+
+	if s.srcConfig.EventMonitoringMethod.Object.isEnabled() {
+		err := s.RunObject()
+		if err != nil {
+			s.log.Errorf("Problem running Object collection: %s", err)
+		}
+	}
+
+	eventLogFileTicker, objectMethodTicker := &time.Ticker{}, &time.Ticker{}
+	eventLogFileTicker.C, objectMethodTicker.C = nil, nil
+
+	if s.srcConfig.EventMonitoringMethod.EventLogFile.isEnabled() {
+		eventLogFileTicker = time.NewTicker(s.srcConfig.EventMonitoringMethod.EventLogFile.Interval)
+		defer eventLogFileTicker.Stop()
+	}
+
+	if s.srcConfig.EventMonitoringMethod.Object.isEnabled() {
+		objectMethodTicker = time.NewTicker(s.srcConfig.EventMonitoringMethod.Object.Interval)
+		defer objectMethodTicker.Stop()
+	}
+
+	for {
+		// Always check for cancel first, to not accidentally trigger another
+		// run if the context is already cancelled, but we have already received
+		// another ticker making the channel ready.
+		select {
```

Comment:
isn't this redundant since the first case in the subsequent select statement is also `s.ctx.Done()`?

Reply:
i've seen a simliar pattern before when an initial collection is triggered without waiting for the first ticker period to complete, but also ensuring that the context is not cancelled first. but in that case the check is outside the loop. in this case, i don't think having it run every time at the start of the loop has any effect, other than checking it twice.

Reply:
Yeah, I know it is confusing. But how "select" works is:

If multiple channels are ready, one is selected at random. So when the context is canceled, etc. such that ctx.Done is ready, it might happen it still goes to another case because that case is also ready; the selection is random.

https://go.dev/tour/concurrency/5

> A select blocks until one of its cases can run, then it executes that case. It chooses one at random if multiple are ready.

So, in some other places in beats too this pattern is followed so that ctx.Done is honored first.

See: https://github.com/elastic/go-concert/blob/b7c3a9ce59e763ea7fb4a7e279c2708950156245/timed/timed.go#L67 (similar)

In beats codebase , you'll find usage of `timed.Periodic` in inputs like httpjson, cel, awss3

Reply:
got it - thanks.